src/libraries/System.Text.Json/Common/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,114 @@
+
+                    if (char.IsLetter(previous) && char.IsLetter(next))
+                    {
+                        // in the middle of a bit of text
+                        var previousUpper = char.IsUpper(previous);
+                        var currentUpper = char.IsUpper(current);
+                        var nextUpper = char.IsUpper(next);
+
+                        switch ((previousUpper, currentUpper, nextUpper))
+                        {
+                            case (false, false, false): // aaa
+                            case ( true,  true,  true): // AAA
+                            case ( true, false, false): // Aaa
+                            {
+                                // same word
+                                result.Append(char.ToLowerInvariant(current));
+                                underlinedAtEndOfWord = false;
+                                break;
+                            }
+
+                            case (false, false,  true): // aaA
+                            case ( true, false,  true): // AaA
+                            {
+                                // end of word
+                                result.Append(char.ToLowerInvariant(current));
+                                result.Append('_');
+                                underlinedAtEndOfWord = true;
+                                break;
+                            }
+
+                            case (false,  true,  true): // aAA
+                            case ( true,  true, false): // AAa
+                            case (false,  true, false): // aAa
+                            {
+                                // beginning of word
+                                if (!underlinedAtEndOfWord)
+                                {
+                                    result.Append('_');
+                                }
+                                result.Append(char.ToLowerInvariant(current));
+                                underlinedAtEndOfWord = false;
+                                break;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // beginning or end of text
+                        result.Append(char.ToLowerInvariant(current));
+                        underlinedAtEndOfWord = false;
+                    }
+                }
+                else if (char.IsLetter(current))
+                {
+                    // text at the beginning or the end of the string
+                    result.Append(char.ToLowerInvariant(current));
+                }
+                else if (char.IsNumber(current))
+                {
+                    // a number at any point in the string
+                    if (x > 0)
+                    {
+                        result.Append('_');
+                    }
+
+                    result.Append(current);
+
+                    if (x < name.Length - 1)
```

Comment:
Thinking about it more, this condition will likely accidentally trigger if two numbers are hit in sequence. In this scenario, what would a user expect the name to be converted to? I know `abc_1__2_def` is most likely incorrect, but should it be `abc_12_def`, or `abc_1_2_def`?

Reply:
For now I've just left this as `abc_1_2_def`, and have added a relevant test case.